internal/encoding/codec.go

```diff
@@ -0,0 +1,135 @@
+package encoding
+
+import (
+	"sync"
+
+	"github.com/spf13/viper/internal/encoding/constructor"
+	"github.com/spf13/viper/internal/encoding/dotenv"
+	"github.com/spf13/viper/internal/encoding/hcl"
+	"github.com/spf13/viper/internal/encoding/ini"
+	"github.com/spf13/viper/internal/encoding/javaproperties"
+	"github.com/spf13/viper/internal/encoding/json"
+	"github.com/spf13/viper/internal/encoding/toml"
+	"github.com/spf13/viper/internal/encoding/yaml"
+)
+
+type Codec interface {
+	// Decode decodes the contents of b into v.
+	// It's primarily used for decoding contents of a file into a map[string]interface{}.
+	Decode(b []byte, v map[string]interface{}) error
+
+	// Encode encodes the contents of v into a byte representation.
+	// It's primarily used for encoding a map[string]interface{} into a file format.
+	Encode(v map[string]interface{}) ([]byte, error)
+
+	// Construct creates an empty struct based on its own type. It is designed to avoid using
+	// reflection in lazy initialization.
+	// constructor.Codec is used to solve the cycle importing.
+	Construct() constructor.Codec
+}
+
+const (
+	// ErrCodecNotFound is returned when there is no codec registered for a format.
+	ErrCodecNotFound = encodingError("codec not found for this format")
+
+	// ErrCodecFormatAlreadyRegistered is returned when a codec is already registered for a format.
+	ErrCodecFormatAlreadyRegistered = encodingError("codec already registered for this format")
+)
+
+// supportedCodecFormats stores all supported codec, the empty pointers are used to construct a corresponding
+// codec object without reflection.
+var supportedCodecFormats = map[string]Codec{
+	"yaml":       (*yaml.Codec)(nil),
+	"yml":        (*yaml.Codec)(nil),
+	"json":       (*json.Codec)(nil),
+	"toml":       (*toml.Codec)(nil),
+	"hcl":        (*hcl.Codec)(nil),
+	"tfvars":     (*hcl.Codec)(nil),
+	"ini":        (*ini.Codec)(nil),
+	"properties": (*javaproperties.Codec)(nil),
+	"props":      (*javaproperties.Codec)(nil),
+	"prop":       (*javaproperties.Codec)(nil),
+	"dotenv":     (*dotenv.Codec)(nil),
+	"env":        (*dotenv.Codec)(nil),
+}
```

Comment:
For the code here, I could use a map to store all constructors, which is more go style. However, this way is slightly faster. I prefer this way.

My benchmark code i listed below and it's inside the `viper/internal/encoding/codec_bench_test.go`.

```go
package encoding

import (
	"testing"

	"github.com/spf13/viper/internal/encoding/yaml"
)

func BenchmarkNilType(b *testing.B) {
	m := map[string]Codec{
		"test": (*yaml.Codec)(nil),
	}
	c := m["test"]
	for i := 0; i < b.N; i++ {
		c.Construct()
	}
}
func BenchmarkWithMapFn(b *testing.B) {
	m := map[string]func() Codec{
		"test": func() Codec {
			return &yaml.Codec{}
		},
	}
	fn := m["test"]
	for i := 0; i < b.N; i++ {
		fn()
	}
}
```

## bench result
```
goos: darwin
goarch: amd64
pkg: github.com/spf13/viper/internal/encoding
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkNilType
BenchmarkNilType-12      	693284240	         1.484 ns/op
BenchmarkWithMapFn
BenchmarkWithMapFn-12    	789944808	         1.528 ns/op
PASS
```